### PR TITLE
Fixes PoV over-estimation

### LINKED ISSUE #13765: [benchmark-cli] Wrong proof_size calculation leading to exaggerated weights
> This has been found during investigation of [x5 proof_size weight](https://github.com/paritytech/substrate/pull/13268/files/50036312437c50c5b491ae49294a7b271cb3bb0a#r1090960965) for one of the pallet_contracts API functions.
This bug was introduced in  https://github.com/paritytech/substrate/pull/11637.

## Context

Results of benchmarks are processed with [`writer::process_storage_results()`](https://github.com/paritytech/substrate/blob/9c92e4987160a17daa72f79186d981b6fbe5879e/utils/frame/benchmarking-cli/src/pallet/writer.rs#L506) in the following way.

It loops through all the storage keys of all results

https://github.com/paritytech/substrate/blob/9c92e4987160a17daa72f79186d981b6fbe5879e/utils/frame/benchmarking-cli/src/pallet/writer.rs#L549-L550

and multiples each benchmark result for each key, adjusting the result's *proof_size* for each key, as it depends on:
- PoV estimation mode
- single read PoV overhead
- number of reads

https://github.com/paritytech/substrate/blob/9c92e4987160a17daa72f79186d981b6fbe5879e/utils/frame/benchmarking-cli/src/pallet/writer.rs#L628-L632 

We get *storage_per_prefix* data as the result, which originally was used for creating comments with information about the storage keys touched during each benchmark.

## The Bug

In [PR#11637](https://github.com/paritytech/substrate/pull/11637/files#diff-9e23aff4915f04fd032342adfc09c096d87ad2a659cd1a6f15719ccf87d537a7R299) this data started to be used for calculation of the resulting *proof_size* formula into the `weights.rs` . **But in a wrong way**:

**Step 1, (almost right)**. We find average base *proof_size* value and *slope* for each component, by making regression analysis of benchmark results we put to this component in  *storage_per_prefix* (see above).

https://github.com/paritytech/substrate/blob/9c92e4987160a17daa72f79186d981b6fbe5879e/utils/frame/benchmarking-cli/src/pallet/writer.rs#L299-L303

*(This is *almost* right but not totally right, because the values we're making regression analysis on are not per-key benchmark results but originated from benchmark results for the all the keys and then adjusted for a single key, see below for details)*

**Step 2, (wrong)**.  The resulting  *base_calculated_proof_size* and *component_calculated_proof_size* values are calculated as a simple **sum** of the values for all prefixes from the *per_prefix* benchmark results. 

https://github.com/paritytech/substrate/blob/9c92e4987160a17daa72f79186d981b6fbe5879e/utils/frame/benchmarking-cli/src/pallet/writer.rs#L317-L323

But, wait a minute. To recap, the path of these *proof_sizes* is:

1. They first were calculated for **all keys** in each benchmark run.
2. Then we **multiplied them for each key**, adjusted depending on reads overhead for the key.
    *(which is weird because we're adding up overhead of a single key PoV to proof_size of all the keys)*
4. Then we **summed up** those multiplied values.

**And this is wrong**. It leads to exaggerated weights when being put to `weights.rs`:

https://github.com/paritytech/substrate/blob/9c92e4987160a17daa72f79186d981b6fbe5879e/.maintain/frame-weight-template.hbs#L73-L75

## Suggested Fix

**Instead** of a simple sum, those *base_calculated_proof_size* and *component_calculated_proof_size*  should be calculated as averages or medians from all keys.

### PR DESCRIPTION
**Pull request by ggwpez**
Closes https://github.com/paritytech/substrate/issues/13765

>(This is almost right but not totally right, because the values we're making regression analysis on are not per-key benchmark results but originated from benchmark results for the all the keys and then adjusted for a single key, see below for details)

Yep, this is in here as follow-up https://github.com/paritytech/substrate/pull/11637. Currently we dont have all the data, so it tries to extrapolate.

TODOs:
- [ ] Update weights (running on `bm2` right now)

### CODE REVIEW

File: utils/frame/benchmarking-cli/src/pallet/writer.rs [hunk 1]
```diff
@@ -315,12 +316,12 @@ fn get_benchmark_data(
...
 	// Sum up the proof sizes per component
 	for (_, slope, base) in proof_size_per_components.iter() {
-		base_calculated_proof_size += base;
+		base_calculated_proof_size = base_calculated_proof_size.max(*base);
 		for component in slope.iter() {
 			let mut found = false;
 			for used_component in used_calculated_proof_size.iter_mut() {
 				if used_component.name == component.name {
-					used_component.slope += component.slope;
+					used_component.slope = used_component.slope.max(component.slope);
```
Comment (agryaznov) [inline utils/frame/benchmarking-cli/src/pallet/writer.rs:324]: so we just take the steepest slope and the largest base among all keys (which could be from two different keys)

**Review by agryaznov — COMMENTED** [review]
Thanks for the fast-fix! 
This should indeed remove excessive over-estimation in its current state.

> > (This is almost right but not totally right, because the values we're making regression analysis on are not per-key benchmark results but originated from benchmark results for the all the keys and then adjusted for a single key, see below for details)
> 
> Yep, this is in here as follow-up #11637. Currently we dont have all the data, so it tries to extrapolate.

Taking just maximum over all keys is still imprecise because it counts for the pov_overhead added by a single key (okay maybe from two: one for base other for slope), and doesn't count for those from all other keys.  That being said, the number of key reads most likely correlates with complexity parameter(s), hence to make a proper approximation we'd need to benchmark per-key, which is currently not implemented. Is that right?

**Comment by ggwpez** [comment]
>Taking just maximum over all keys is still imprecise because it counts for the pov_overhead added by a single key (okay maybe from two: one for base other for slope), and doesn't count for those from all other keys. That being said, the number of key reads most likely correlates with complexity parameter(s), hence to make a proper approximation we'd need to benchmark per-key, which is currently not implemented. Is that right?

I dont quite get how you mean the first part. We have to count overhead of every key once. So if two keys are read, then we have to count the overhead twice since in the worst case these keys dont share any common nodes in the proof.  

About the second part: yes. Currently we just measure the proof size of the whole thing (per component). It would be better to reset the proof recorder after every access to track the size of each key.  
In the end we want to know how often each key is read in correlation to each complexity parameter. So we need a map of proof sizes like `(KEYS, PARAMS) => POV-SIZE`.

File: utils/frame/benchmarking-cli/src/pallet/writer.rs (hunk 1 above)
Comment (ggwpez) [reply utils/frame/benchmarking-cli/src/pallet/writer.rs:324]: Yea but by what do you want to prioritize? If we go by largest base then it possibly underestimates the slope and vice versa.

**Comment by agryaznov** [comment]
> We have to count overhead of every key once. So if two keys are read, then we have to count the overhead twice since in the worst case these keys dont share any common nodes in the proof.

This is not what happening in the current implementation. We currently calculate the overhead for a key for each benchmark, add it to proof_size of each benchmark, then approximate over those results ending up in base and slope for a single key. Then repeat this for all other keys. Therefore we end up having maximum base among all keys, but it could be larger if it had counted overhead for more than single key. But it hadn't, it counts overhead only for a single key. Same for slopes.

**Comment by ggwpez** [comment]
bot bench $ all

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2664092

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2664093

**Comment by ggwpez** [comment]
@agryaznov the proof size comparison is [here](https://weights.tasty.limo/compare?repo=substrate&threshold=1&path_pattern=frame%2F**%2Fsrc%2Fweights.rs&method=asymptotic&ignore_errors=true&unit=proof&old=6b78a413a2d776384169d771f0f69120b8eb264b&new=oty-fix-pov-bench&git_pull=false).  
And downloaded as PDF
[SWC-Web.pdf](https://github.com/paritytech/substrate/files/11199232/SWC-Web.pdf). The values all went down, also the most offending function now has one byte per `c`:  
image

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2668789

**Review by athei — APPROVED** [review]

**Review by agryaznov — APPROVED** [review]
Looks okay as a temp hotfix until https://github.com/paritytech/substrate/issues/13808 is done.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2674331

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2674317

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2674690

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2674676

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2674675

**Comment by ggwpez** [comment]
bot merge

**Pull request merged**